### PR TITLE
Godocs for core packages

### PR DESCRIPTION
**Pull request by yusefnapora**
This is a start on iteration two of @raulk's master plan outlined in #1 

So far I've added package-level comments and tried to make sure public interfaces and types have at least some kind of doc comment. I also replaced lingering references to IPFS with libp2p in a few places.

The methods on the `Negotiator` interface in `protocol/switch.go` weren't familiar enough for me to take a stab at - I'll dig a little into the implementation to see if I can figure it out. And / or if anyone wants to explain it to me, that works :smiley: 

I based the PR off of the consolidate-skeleton branch from #1 and can keep rebasing as that PR progresses.

Let me know if I've gotten anything wrong along the way!

### CODE REVIEW

File: connmgr/connmgr.go [hunk 1]
```diff
@@ -1,3 +1,9 @@
+// Package connmgr provides connection tracking and management interfaces for libp2p.
+//
+// The ConnManager interface exported from this package allows libp2p to enforce an
+// upper bound on the total number of open connections, which prevents us from
+// overwhelming routers with too many connections. Without such a mechanism, many
+// consumer routers will begin to fail and behave erratically given enough connections.
```
Comment (vyzo) [inline connmgr/connmgr.go]: Well that's not the only reason to enforce upper bound in connections. We also care about memory and resource usage from large number of connections.

File: connmgr/connmgr.go [hunk 2]
```diff
@@ -12,6 +18,9 @@ import (
 // with each peer.
 //
 // It enables connections to be trimmed based on implementation-defined heuristics.
+// The ConnManager allows libp2p to enforce an upper bound on the total number of
+// open connections. Without such a bound, we risk overwhelming and breaking some
+// routers (especially consumer models).
```
Comment (vyzo) [inline connmgr/connmgr.go]: I would deemphasize the killing routers issue here.

File: connmgr/null.go [hunk 3]
```diff
@@ -7,6 +7,8 @@ import (
 	"github.com/libp2p/go-libp2p-core/peer"
 )
 
+// NullConnMgr is a ConnMgr that provides no functionality. It is used by go-libp2p-blankhost
+// to satisfy the ConnMgr interface.
```
Comment (vyzo) [inline connmgr/null.go]: The forward reference to blankhost might be unnecessary. Also, it is the default if we don't specify a `ConnManager` option in the libp2p constructor.

File: connmgr/connmgr.go (hunk 1 above)
Comment (yusefnapora) [reply connmgr/connmgr.go]: That's a good point; it's probably enough just to mention that we want to impose an upper bound. There are enough obvious motivations that we don't need to distract people by talking about choking routers.

File: crypto/rsa.go [hunk 4]
```diff
@@ -61,11 +63,13 @@ func (pk *RsaPublicKey) Bytes() ([]byte, error) {
 	return MarshalPublicKey(pk)
 }
 
+// Raw returns a binary representation of the key in a type-specific format.
+// In the case of RSA keys, returns
```
Comment (Stebalien) [inline crypto/rsa.go]: returns...

An X.509 PKIX encoded public key.

File: crypto/rsa.go [hunk 5]
```diff
@@ -61,11 +63,13 @@ func (pk *RsaPublicKey) Bytes() ([]byte, error) {
...
 }
 
+// Raw returns a binary representation of the key in a type-specific format.
+// In the case of RSA keys, returns 
 func (pk *RsaPublicKey) Raw() ([]byte, error) {
 	return x509.MarshalPKIXPublicKey(pk.k)
 }
 
-// Encrypt returns encrypted bytes from the inpu data
+// Encrypt returns encrypted bytes from the input data
```
Comment (Stebalien) [inline crypto/rsa.go]: .

File: metrics/reporter.go [hunk 6]
```diff
@@ -1,17 +1,21 @@
+// Package metrics provides metrics collection and reporting interfaces for libp2p.
 package metrics
 
 import (
 	"github.com/libp2p/go-libp2p-core/peer"
 	"github.com/libp2p/go-libp2p-core/protocol"
 )
 
+// Stats represents a point-in-time snapshot of bandwidth metrics
+// TODO(yusef): clarify units for Rate fields
```
Comment (Stebalien) [inline metrics/reporter.go]: Bytes and seconds.

Bla bla networking people use bits but I use bytes.

File: protocol/switch.go [hunk 7]
```diff
@@ -6,21 +7,54 @@ import (
...
+// are accepted.
 type Router interface {
+
+	// AddHandler registers the given handler to be invoked for
+	// an exact literal match of the given protocol ID string.
 	AddHandler(protocol string, handler HandlerFunc)
+
+	// AddHandlerWithFunc registers the given handler to be invoked
+	// for exact literal matches of the given protocol ID string,
+	// **or** when the provided match function returns true.
```
Comment (Stebalien) [inline protocol/switch.go]: When a match function is provided, the protocol string is entirely ignored.

File: protocol/switch.go [hunk 8]
```diff
@@ -6,21 +7,54 @@ import (
...
+	//
+	// The match function will be invoked with an incoming protocol
+	// ID string when the router is unable to find an exact literal
+	// match among the registered handlers.
```
Comment (Stebalien) [inline protocol/switch.go]: We just check each protocol name/match function in order or registration.

File: protocol/switch.go [hunk 9]
```diff
@@ -6,21 +7,54 @@ import (
...
+	// included in this list if handlers were added with match functions
+	// using AddHandlerWithFunc.
 	Protocols() []string
 }
 
+// Negotiator is a component capable of reaching agreement over what protocols
+// to use for a given stream of communication.
+//
+// The Negotiator is responsible for proposing the protocol to use for a given
+// stream, and it responds to proposals sent by the other end of the stream.
```
Comment (Stebalien) [inline protocol/switch.go]: The negotiator just responds to proposals, it won't propose anything. To propose, one uses one of the `Select*` functions.

File: protocol/switch.go [hunk 10]
```diff
@@ -6,21 +7,54 @@ import (
...
 type Negotiator interface {
 	NegotiateLazy(rwc io.ReadWriteCloser) (io.ReadWriteCloser, string, HandlerFunc, error)
```
Comment (Stebalien) [inline protocol/switch.go]: Negotiate an inbound connection but returns as soon as we learn the protocol for the connection before we finish responding.

(I'm not actually sure if this is useful in practice but we can reconsider it in round 2)

File: protocol/switch.go [hunk 11]
```diff
@@ -6,21 +7,54 @@ import (
...
 	Negotiate(rwc io.ReadWriteCloser) (string, HandlerFunc, error)
```
Comment (Stebalien) [inline protocol/switch.go]: Picks one of the protocols registered with `AddHandler` and friends and returns the protocol name and registered handler function.

File: protocol/switch.go [hunk 12]
```diff
@@ -6,21 +7,54 @@ import (
...
 	Handle(rwc io.ReadWriteCloser) error
```
Comment (Stebalien) [inline protocol/switch.go]: Calls Negotiate and then calls the registered handler function (returns an error if negotiation fails).

File: routing/routing.go [hunk 13]
```diff
@@ -20,6 +20,9 @@ var ErrNotSupported = errors.New("routing: operation or key not supported")
 
 // ContentRouting is a value provider layer of indirection. It is used to find
 // information about who has what content.
+//
+// Content is identified by Cid (short for content ID), which encodes a hash
```
Comment (Stebalien) [inline routing/routing.go]: We usually capitalize CID.

File: crypto/rsa.go [hunk 14]
```diff
@@ -98,22 +102,24 @@ func (sk *RsaPrivateKey) Type() pb.KeyType {
...
 }
 
-// Equals checks whether this key is equal to another
+// Equals checks whether this key is equal to another.
 func (sk *RsaPrivateKey) Equals(k Key) bool {
 	return KeyEqual(sk, k)
 }
 
-// UnmarshalRsaPrivateKey returns a private key from the input x509 bytes
+// UnmarshalRsaPrivateKey returns a private key from the input x509 bytes.
```
Comment (raulk) [inline crypto/rsa.go]:
```suggestion
// UnmarshalRsaPrivateKey returns a private key from the input X.509 bytes.
```

File: crypto/rsa.go [hunk 15]
```diff
@@ -125,12 +131,12 @@ func UnmarshalRsaPrivateKey(b []byte) (PrivKey, error) {
 	return &RsaPrivateKey{sk: sk}, nil
 }
 
-// MarshalRsaPrivateKey returns the x509 bytes of the private key
+// MarshalRsaPrivateKey returns the x509 bytes of the private key.
```
Comment (raulk) [inline crypto/rsa.go]:
```suggestion
// MarshalRsaPrivateKey returns the X.509 bytes of the private key.
```

File: crypto/rsa.go [hunk 16]
```diff
@@ -125,12 +131,12 @@ func UnmarshalRsaPrivateKey(b []byte) (PrivKey, error) {
...
 func MarshalRsaPrivateKey(k *RsaPrivateKey) []byte {
 	return x509.MarshalPKCS1PrivateKey(k.sk)
 }
 
-// UnmarshalRsaPublicKey returns a public key from the input x509 bytes
+// UnmarshalRsaPublicKey returns a public key from the input x509 bytes.
```
Comment (raulk) [inline crypto/rsa.go]:
```suggestion
// UnmarshalRsaPublicKey returns a public key from the input X.509 bytes.
```

File: crypto/rsa.go [hunk 17]
```diff
@@ -146,7 +152,7 @@ func UnmarshalRsaPublicKey(b []byte) (PubKey, error) {
 	return &RsaPublicKey{pk}, nil
 }
 
-// MarshalRsaPublicKey returns the x509 bytes from the public key
+// MarshalRsaPublicKey returns the x509 bytes from the public key.
```
Comment (raulk) [inline crypto/rsa.go]:
```suggestion
// MarshalRsaPublicKey returns the X.509 bytes from the public key.
```

File: host/host.go [hunk 18]
```diff
@@ -1,3 +1,6 @@
+// Package host provides the core Host interface for libp2p.
+//
+// Host represents a single libp2p "agent" that can act as client or server.
```
Comment (raulk) [inline host/host.go]: Client/server terminology is a wee bit confusing in p2p.

```suggestion
// Host represents a single libp2p node in a peer-to-peer network.
```

File: metrics/bandwidth.go [hunk 19]
```diff
@@ -17,28 +21,39 @@ type BandwidthCounter struct {
 	peerOut flow.MeterRegistry
 }
 
+// NewBandwidthCounter creates a new BandwidthCounter
```
Comment (raulk) [inline metrics/bandwidth.go]:
```suggestion
// NewBandwidthCounter creates a new BandwidthCounter.
```

File: metrics/bandwidth.go [hunk 20]
```diff
@@ -17,28 +21,39 @@ type BandwidthCounter struct {
...
 }
 
+// LogRecvMessage records the size of an incoming message
+// without associating the bandwith to a specific peer or protocol.
 func (bwc *BandwidthCounter) LogRecvMessage(size int64) {
 	bwc.totalIn.Mark(uint64(size))
 }
 
+// LogSentMessageStream records the size of an outgoing message over a single logical stream.
+// Bandwidth is associated with the given protocol.ID and peer.ID
```
Comment (raulk) [inline metrics/bandwidth.go]:
```suggestion
// Bandwidth is associated with the given protocol.ID and peer.ID.
```

File: metrics/bandwidth.go [hunk 21]
```diff
@@ -17,28 +21,39 @@ type BandwidthCounter struct {
...
 func (bwc *BandwidthCounter) LogSentMessageStream(size int64, proto protocol.ID, p peer.ID) {
 	bwc.protocolOut.Get(string(proto)).Mark(uint64(size))
 	bwc.peerOut.Get(string(p)).Mark(uint64(size))
 }
 
+// LogRecvMessageStream records the size of an incoming message over a single logical stream.
+// Bandwidth is associated with the given protocol.ID and peer.ID
```
Comment (raulk) [inline metrics/bandwidth.go]:
```suggestion
// Bandwidth is associated with the given protocol.ID and peer.ID.
```

File: metrics/reporter.go [hunk 22]
```diff
@@ -1,17 +1,23 @@
...
 package metrics
 
 import (
 	"github.com/libp2p/go-libp2p-core/peer"
 	"github.com/libp2p/go-libp2p-core/protocol"
 )
 
+// Stats represents a point-in-time snapshot of bandwidth metrics.
+//
+// The TotalIn and TotalOut fields record bytes sent / received to date.
```
Comment (raulk) [inline metrics/reporter.go]:
```suggestion
// The TotalIn and TotalOut fields record cumulative bytes sent / received.
```

File: network/errors.go [hunk 23]
```diff
@@ -2,7 +2,7 @@ package network
 
 import "errors"
 
-// There are no addresses associated with a peer when they were needed.
+// ErrNoRemoteAddrs is returned when there are no addresses associated with a peer when they were needed.
```
Comment (raulk) [inline network/errors.go]:
```suggestion
// ErrNoRemoteAddrs is returned when there are no addresses associated with a peer during a dial.
```

File: mux/mux.go [hunk 24]
```diff
@@ -28,10 +32,19 @@ type MuxedStream interface {
...
 
-// MuxedConn is a stream-multiplexing connection to a remote peer.
+// MuxedConn represents a connection to a remote peer that has been
+// extended to support stream multiplexing.
+//
+// A MuxedConn allows a single net.Conn connection to carry many logically
+// independent bidirectional streams of binary data.
+//
+// Together with network.ConnSecurity, MuxedConn is a component of the
+// transport.UpgradedConn interface, which represents a "raw" network
```
Comment (raulk) [inline mux/mux.go]:
```suggestion
// transport.CapableConn interface, which represents a "raw" network
```

File: mux/mux.go [hunk 25]
```diff
@@ -28,10 +32,19 @@ type MuxedStream interface {
...
+// connection that has been "upgraded" to support the libp2p "table stakes"
```
Comment (raulk) [inline mux/mux.go]:
```suggestion
// connection that has been "upgraded" to support the libp2p capabilities
```

File: mux/mux.go [hunk 26]
```diff
@@ -47,7 +60,9 @@ type MuxedConn interface {
 	AcceptStream() (MuxedStream, error)
 }
 
-// Transport constructs go-stream-muxer compatible connections.
+// Multiplexer "wraps" a net.Conn with a stream multiplexing
```
Comment (raulk) [inline mux/mux.go]:
```suggestion
// Multiplexer wraps a net.Conn with a stream multiplexing
```

File: network/network.go [hunk 27]
```diff
@@ -1,3 +1,8 @@
+// Package network provides core networking abstractions for libp2p.
+//
+// The network package provides the high-level Network interface for interacting
+// with other libp2p peers, which is the primary public API for initiating and
+// accepting connections to remote peers.
```
Comment (raulk) [inline network/network.go]: The Network is generally a confusing component for users. We'll need to expand on this description. Possible content: The Network is the component that interacts with the underlying physical network stack in order to establish outbound connections and accept inbound connections to/from peers.

File: peer/peer.go [hunk 28]
```diff
@@ -34,14 +34,17 @@ var AdvancedEnableInlining = true
 const maxInlineKeyLength = 42
 
 // ID is a libp2p peer identity.
+//
+// Peer IDs are derived by hashing a peer's public key and encoding the
+// hash output as a multihash. See IDFromPublicKey for details.
 type ID string
 
-// Pretty returns a b58-encoded string of the ID
+// Pretty returns a base58-encoded string representation of the ID
```
Comment (raulk) [inline peer/peer.go]:
```suggestion
// Pretty returns a base58-encoded string representation of the ID.
```

File: peer/peer.go [hunk 29]
```diff
@@ -34,14 +34,17 @@ var AdvancedEnableInlining = true
...
 func (id ID) Pretty() string {
 	return IDB58Encode(id)
 }
 
-// Loggable returns a pretty peerID string in loggable JSON format
+// Loggable returns a pretty peer ID string in loggable JSON format
```
Comment (raulk) [inline peer/peer.go]:
```suggestion
// Loggable returns a pretty peer ID string in loggable JSON format.
```

File: peer/peer.go [hunk 30]
```diff
@@ -66,12 +69,12 @@ func (id ID) ShortString() string {
 	return fmt.Sprintf("<peer.ID %s*%s>", pid[:2], pid[len(pid)-6:])
 }
 
-// MatchesPrivateKey tests whether this ID was derived from sk
+// MatchesPrivateKey tests whether this ID was derived from the secret key sk
```
Comment (raulk) [inline peer/peer.go]:
```suggestion
// MatchesPrivateKey tests whether this ID was derived from the secret key sk.
```

File: peer/peer.go [hunk 31]
```diff
@@ -66,12 +69,12 @@ func (id ID) ShortString() string {
...
 func (id ID) MatchesPrivateKey(sk ic.PrivKey) bool {
 	return id.MatchesPublicKey(sk.GetPublic())
 }
 
-// MatchesPublicKey tests whether this ID was derived from pk
+// MatchesPublicKey tests whether this ID was derived from the public key pk
```
Comment (raulk) [inline peer/peer.go]:
```suggestion
// MatchesPublicKey tests whether this ID was derived from the public key pk.
```

File: peer/peer.go [hunk 32]
```diff
@@ -99,7 +102,7 @@ func (id ID) ExtractPublicKey() (ic.PubKey, error) {
 	return pk, nil
 }
 
-// Validate check if ID is empty or not
+// Validate checks if ID is empty or not
```
Comment (raulk) [inline peer/peer.go]:
```suggestion
// Validate checks if ID is empty or not.
```

File: peer/peer.go [hunk 33]
```diff
@@ -108,7 +111,7 @@ func (id ID) Validate() error {
 	return nil
 }
 
-// IDFromString cast a string to ID type, and validate
+// IDFromString casts a string to ID type, and validates
 // the id to make sure it is a multihash.
```
Comment (raulk) [inline peer/peer.go]:
```suggestion
// the value to make sure it is a multihash.
```

File: peer/peer.go [hunk 34]
```diff
@@ -117,7 +120,7 @@ func IDFromString(s string) (ID, error) {
 	return ID(s), nil
 }
 
-// IDFromBytes cast a string to ID type, and validate
+// IDFromBytes casts a byte slice to ID type, and validates
```
Comment (raulk) [inline peer/peer.go]:
```suggestion
// IDFromBytes casts a byte slice to the ID type, and validates
```

File: peer/peer.go [hunk 35]
```diff
@@ -108,7 +111,7 @@ func (id ID) Validate() error {
 	return nil
 }
 
-// IDFromString cast a string to ID type, and validate
+// IDFromString casts a string to ID type, and validates
```
Comment (raulk) [inline peer/peer.go]:
```suggestion
// IDFromString casts a string to the ID type, and validates
```

File: peer/peer.go [hunk 36]
```diff
@@ -117,7 +120,7 @@ func IDFromString(s string) (ID, error) {
 	return ID(s), nil
 }
 
-// IDFromBytes cast a string to ID type, and validate
+// IDFromBytes casts a byte slice to ID type, and validates
 // the id to make sure it is a multihash.
```
Comment (raulk) [inline peer/peer.go]:
```suggestion
// the value to make sure it is a multihash.
```

File: peer/peer.go [hunk 37]
```diff
@@ -149,12 +154,12 @@ func IDHexDecode(s string) (ID, error) {
...
 }
 
-// IDHexEncode returns hex-encoded string
+// IDHexEncode returns the hex-encoded multihash representation of the ID.
 func IDHexEncode(id ID) string {
 	return hex.EncodeToString([]byte(id))
 }
 
-// IDFromPublicKey returns the Peer ID corresponding to pk
+// IDFromPublicKey returns the Peer ID corresponding to the public key pk
```
Comment (raulk) [inline peer/peer.go]:
```suggestion
// IDFromPublicKey returns the Peer ID corresponding to the public key pk.
```

File: peer/peer.go [hunk 38]
```diff
@@ -168,7 +173,7 @@ func IDFromPublicKey(pk ic.PubKey) (ID, error) {
 	return ID(hash), nil
 }
 
-// IDFromPrivateKey returns the Peer ID corresponding to sk
+// IDFromPrivateKey returns the Peer ID corresponding to the secret key sk
```
Comment (raulk) [inline peer/peer.go]:
```suggestion
// IDFromPrivateKey returns the Peer ID corresponding to the secret key sk.
```

File: peerstore/peerstore.go [hunk 39]
```diff
@@ -1,3 +1,4 @@
+// Package peerstore provides types and interfaces for local storage of metadata about libp2p peers.
```
Comment (raulk) [inline peerstore/peerstore.go]:
```suggestion
// Package peerstore provides types and interfaces for local storage of address information, 
// metadata, and public key material about libp2p peers.
```

File: sec/insecure/insecure.go [hunk 40]
```diff
@@ -15,7 +18,8 @@ import (
 const ID = "/plaintext/1.0.0"
 
 // Transport is a no-op stream security transport. It provides no
-// security and simply wraps connections in blank
+// security and simply wraps connections with the secure transport
```
Comment (raulk) [inline sec/insecure/insecure.go]:
```suggestion
// security and simply mocks the security and identity methods to
// return peer IDs known ahead of time.
```

File: sec/insecure/insecure.go [hunk 41]
```diff
@@ -15,7 +18,8 @@ import (
...
+// interface.
```
Comment (raulk) [inline sec/insecure/insecure.go]:
```suggestion
```

File: transport/transport.go [hunk 42]
```diff
@@ -34,9 +39,16 @@ type UpgradedConn interface {
...
-// connections from other peers. The built-in transports provided are TCP and UTP
-// but many more can be implemented, sctp, audio signals, sneakernet, UDT, a
-// network of drones carrying usb flash drives, and so on.
+// connections from other peers.
+//
+// The Transport interface allows you to open connections to other peers
+// by dialing them, and also lets you listen for incoming connections.
+//
+// Connections returned by Dial and passed into Listeners are of type
+// UpgradedConn, which means that they have been upgraded to support
```
Comment (raulk) [inline transport/transport.go]:
```suggestion
// CapableConn, which means that they have been upgraded to support
```

File: transport/transport.go [hunk 43]
```diff
@@ -67,7 +79,7 @@ type Transport interface {
 
 // Listener is an interface closely resembling the net.Listener interface.  The
```
Comment (raulk) [inline transport/transport.go]:
```suggestion
// Listener is an interface closely resembling the net.Listener interface. The
```

File: transport/transport.go [hunk 44]
```diff
@@ -67,7 +79,7 @@ type Transport interface {
...
 // only real difference is that Accept() returns Conn's of the type in this
-// package, and also exposes a Multiaddr method as opposed to a regular Addr
+// package, and also exposes a Multiaddr method as well as a regular Addr
```
Comment (raulk) [inline transport/transport.go]: I think the original doc was correct.

```suggestion
// package, and also exposes a Multiaddr method as opposed to a regular Addr
```

**Comment by raulk** [comment]
In the interest of pushing this forward, I'm gonna accept my own suggestions, but I'll leave the comments unresolved so you can review them @yusefnapora and give me feedback. I'm also merging from the target branch to resolve conflicts (esp. UpgradedConn => CapableConn rename).

**Pull request merged**